pyresample/slicer.py

```diff
@@ -0,0 +1,213 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2021-2022 Pyresample developers
+#
+# This program is free software: you can redistribute it and/or modify it under
+# the terms of the GNU Lesser General Public License as published by the Free
+# Software Foundation, either version 3 of the License, or (at your option) any
+# later version.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+# details.
+#
+# You should have received a copy of the GNU Lesser General Public License along
+# with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Area and Swath Slicers."""
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from functools import lru_cache
+
+import numpy as np
+from pyproj import Transformer
+
+from pyresample import AreaDefinition, SwathDefinition
+from pyresample.geometry import (
+    IncompatibleAreas,
+    InvalidArea,
+    get_geostationary_bounding_box_in_proj_coords,
+)
+
+try:
+    import dask.array as da
+except ImportError:
+    da = None
+
+
+def create_slicer(area_to_crop, area_to_contain):
+    """Create a slicer for cropping *area_to_crop* based on *area_to_contain*.
+
+    Return an AreaSlicer or a SwathSlicer based on the first area type.
+    """
+    if isinstance(area_to_crop, SwathDefinition):
+        return SwathSlicer(area_to_crop, area_to_contain)
+    elif isinstance(area_to_crop, AreaDefinition):
+        return AreaSlicer(area_to_crop, area_to_contain)
+    else:
+        raise NotImplementedError("Don't know how to slice a " + str(type(area_to_crop)))
+
+
+class Slicer(ABC):
+    """Abstract Slicer.
+
+    Provided an Area-to-crop and an Area-to-contain, a Slicer provides methods
+    to find slices that enclose `area-to-contain` inside `area-to-crop`.
+    """
```

Comment:
It may just be late here, but would it be possible to provide an example of this? Like "For example, a full disk geostationary AreaDefinition be sliced to a Lambert Conformal Conic projection projection over Germany"...?

Reply:
never too late :D

Reply:
Looks good. Thanks.